distributed/scheduler.py

```diff
@@ -7488,36 +7532,92 @@ def decide_worker(
     of bytes sent between workers.  This is determined by calling the
     *objective* function.
     """
-    ws: WorkerState = None
     wws: WorkerState
-    dts: TaskState
+
+    group: TaskGroup = ts._group
+    ws: WorkerState = group._last_worker
+
+    if valid_workers is not None:
+        total_nthreads = sum(wws._nthreads for wws in valid_workers)
```

Comment:
This walks through all workers for all tasks.  We may not be able to do this.

Reply:
See below; I believe `valid_workers is None` is the common case? Agreed that this isn't ideal though. But if there are worker restrictions, ignoring them and just using `self._total_nthreads` could be wildly wrong (imagine 10 GPU workers and 100 CPU workers for a task group of 50 that needs to run on GPUs). Maybe there's a cheaper measurement?

Reply:
OK, grand.